src/librustc/ty/layout.rs

```diff
@@ -1025,65 +1099,27 @@ impl<'tcx> LayoutCx<'tcx, TyCtxt<'tcx>> {
                     }
                 }
 
-                let (mut min, mut max) = (i128::max_value(), i128::min_value());
-                let discr_type = def.repr.discr_type();
-                let bits = Integer::from_attr(self, discr_type).size().bits();
-                for (i, discr) in def.discriminants(tcx) {
-                    if variants[i].iter().any(|f| f.abi.is_uninhabited()) {
-                        continue;
-                    }
-                    let mut x = discr.val as i128;
-                    if discr_type.is_signed() {
-                        // sign extend the raw representation to be an i128
-                        x = (x << (128 - bits)) >> (128 - bits);
-                    }
-                    if x < min { min = x; }
-                    if x > max { max = x; }
-                }
-                // We might have no inhabited variants, so pretend there's at least one.
-                if (min, max) == (i128::max_value(), i128::min_value()) {
-                    min = 0;
-                    max = 0;
-                }
-                assert!(min <= max, "discriminant range is {}...{}", min, max);
-                let (min_ity, signed) = Integer::repr_discr(tcx, ty, &def.repr, min, max);
-
-                let mut align = dl.aggregate_align;
-                let mut size = Size::ZERO;
-
-                // We're interested in the smallest alignment, so start large.
-                let mut start_align = Align::from_bytes(256).unwrap();
-                assert_eq!(Integer::for_align(dl, start_align), None);
+                let mut prefix_align = ity.align(dl).abi;
+                let align = align.max(AbiAndPrefAlign::new(prefix_align));
 
                 // repr(C) on an enum tells us to make a (tag, union) layout,
                 // so we need to grow the prefix alignment to be at least
                 // the alignment of the union. (This value is used both for
                 // determining the alignment of the overall enum, and the
                 // determining the alignment of the payload after the tag.)
-                let mut prefix_align = min_ity.align(dl).abi;
                 if def.repr.c() {
-                    for fields in &variants {
-                        for field in fields {
-                            prefix_align = prefix_align.max(field.align.abi);
-                        }
-                    }
+                    prefix_align = align.abi;
```

Comment:
The fact that some of this code was moved up but not all of it makes me uneasy. I can't easily tell if it will always produce the same result.

Reply:
The niche filling code needs to use the alignment without `ity.align(dl)`. I suppose I could use a different name for that, maybe `body_align`, and only set `align` after we know that we'll use a tag.

Reply:
Nevermind, seems to be #63902.

Reply:
I would very much prefer to fix the bug without landing #63902.